pkg/ring/ring.go

```diff
@@ -326,16 +326,37 @@ func (r *Ring) GetAll(op Operation) (ReplicationSet, error) {
 		return ReplicationSet{}, ErrEmptyRing
 	}
 
+	var maxUnavailableZones int
+	var numRequired int
+	ingesters := make([]IngesterDesc, 0, len(r.ringDesc.Ingesters))
+	if r.cfg.ZoneAwarenessEnabled {
+		zones := make(map[string]bool)
+
+		for _, ingester := range r.ringDesc.Ingesters {
+			zones[ingester.Zone] = true
+			ingesters = append(ingesters, ingester)
+		}
+		maxUnavailableZones = len(zones) / 2
+		// Assume ingesters are evenly spread over zones
+		// The maximum number of failures is all the ingesters in the maximum number of failing zones
+		maxErrors := (len(ingesters) / len(zones)) * maxUnavailableZones
```

Comment:
No, this is not correct. With `MaxErrors` you're not telling "max number of ingesters failing in 1 specific zone", but you're telling "max number of ingesters failing across the whole cluster". When zone awareness is enabled, max errors is... it depends:

1. If no instance has been filtered out because unhealthy, then it's equal to `maxUnavailable` as computed below (when zone-awareness is disabled)
2. If instances in 1 zone have been all filtered out (which should be when the number of failing instances in a single zone are > `maxUnavailable`) then `MaxErrors: 0` (because we can't sustain any other failure)

Reply:
The cases should now be handled as well.